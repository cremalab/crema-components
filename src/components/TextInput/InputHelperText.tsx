import { ReactNode } from "react"

interface Props {
  helperText: ReactNode
  helperTextId: string
}

export function InputHelperText({ helperTextId, helperText }: Props) {
  return <div id={helperTextId}>{helperText}</div>
}
