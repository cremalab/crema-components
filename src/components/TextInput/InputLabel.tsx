interface Props {
  className?: string
  label: string
  inputLabelId: string
  uid: string
}

export function InputLabel({ className, uid, inputLabelId, label }: Props) {
  return (
    <label id={inputLabelId} htmlFor={uid} className={className}>
      {label}
    </label>
  )
}
