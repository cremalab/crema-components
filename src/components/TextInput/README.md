# `<TextInput />`

DESCRIPTION_HERE

## Directory Structure

- `TextInput.stories.tsx`: Component stories (`npm run test:playground`)
- `TextInput.test.tsx`: Component tests (`npm run test:unit`)
- `TextInput.tsx`: Component code
- `index.ts`: Component export
- `README.md`: Component documentation (hey, that's me!)

## Example

![TextInput](../../../.loki/reference/chrome_TextInput_example.png)

> Image location: [`.loki/reference/chrome_TextInput_example.png`](../../../.loki/reference/chrome_TextInput_example.png)
> 
> To regenerate: 
> 1. `npm run test:playground` (skip if running)
> 1. `npm run test:visual:update -- --storiesFilter="^TextInput example\$"`