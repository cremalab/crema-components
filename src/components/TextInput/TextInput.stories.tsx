import { ComponentMeta, Story } from "@storybook/react"
import { ComponentProps } from "react"
import { TextInput } from "."

/**
 * See Storybook Docs: Writing Stories
 * https://storybook.js.org/docs/basics/writing-stories/
 */

export default {
  title: "Components/TextInput",
  component: TextInput,
} as ComponentMeta<typeof TextInput>

const Template: Story<ComponentProps<typeof TextInput>> = (args) => (
  <TextInput {...args} />
)

export const Default = Template.bind({})

Default.args = {
  className: "",
  disabled: false,
  helperText: "Enter some text",
  hideLabel: false,
  id: "1",
  label: "Base Input",
  name: "Base Input",
  type: "text",
}

export const Disabled = Template.bind({})

Disabled.args = {
  className: "",
  disabled: true,
  helperText: "Enter some text",
  hideLabel: false,
  id: "1",
  label: "Disabled Input",
  name: "Disabled Input",
  type: "text",
}

export const LabelHidden = Template.bind({})

LabelHidden.args = {
  className: "",
  disabled: true,
  helperText: "Enter some text",
  hideLabel: true,
  id: "1",
  label: "LabelHidden Input",
  name: "LabelHidden Input",
  type: "text",
}

export const WithPlaceholder = Template.bind({})

WithPlaceholder.args = {
  className: "",
  disabled: false,
  hideLabel: false,
  id: "1",
  label: "Placeholder Input",
  name: "Placeholder Input",
  placeholder: "Placeholder text",
  type: "text",
}

export const WithHelperText = Template.bind({})

WithHelperText.args = {
  className: "",
  disabled: false,
  hideLabel: false,
  id: "1",
  label: "Helper Text Input",
  name: "Helper Text Input",
  placeholder: "Placeholder text",
  type: "text",
  helperText: "Helper text",
}
