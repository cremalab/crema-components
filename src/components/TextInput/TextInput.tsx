import { HTMLProps, ReactNode, Ref, forwardRef } from "react"
import { InputHelperText } from "./InputHelperText"
import { InputLabel } from "./InputLabel"

interface TextInputProps extends HTMLProps<HTMLInputElement> {
  helperText?: ReactNode
  hideLabel?: boolean
  inputRef?: Ref<HTMLInputElement>
  name: string
}

export const TextInput = forwardRef<HTMLDivElement, TextInputProps>(
  function TextInput(
    { helperText, hideLabel, inputRef, name, ...inputProps }: TextInputProps,
    ref,
  ) {
    const uid = name
    const inputLabelId = `${uid}-label`
    const helperTextId = `${uid}-helper-text`

    const inputAriaAttributes: Record<string, unknown> = {}

    inputAriaAttributes["aria-label"] = inputProps?.label

    if (helperText) {
      inputAriaAttributes["aria-describedby"] = helperTextId
    }

    return (
      <div ref={ref}>
        {inputProps?.label && !hideLabel ? (
          <InputLabel
            label={inputProps?.label}
            uid={uid}
            inputLabelId={inputLabelId}
          />
        ) : null}
        <div>
          <input
            name={name}
            ref={inputRef}
            {...inputAriaAttributes}
            {...inputProps}
          />
        </div>
        {helperText && (
          <InputHelperText
            helperText={helperText}
            helperTextId={helperTextId}
          />
        )}
      </div>
    )
  },
)
