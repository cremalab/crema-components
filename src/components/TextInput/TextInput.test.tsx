import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { TextInput } from "./TextInput"

describe("TextInput", () => {
  it("renders the expected markup", () => {
    const { asFragment } = render(
      <TextInput
        helperText="Test Helper Text"
        name="test-input"
        label="Test Input"
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it("renders an input w/ the required attributes", async () => {
    render(<TextInput name="test-input" label="Test Input" />)
    const received = await screen.findByLabelText("Test Input")

    expect(received).toBeInTheDocument()
    expect(received).toHaveAttribute("name", "test-input")
    expect(received).toHaveAttribute("aria-label", "Test Input")
  })

  it("input calls the most used events during usage", async () => {
    const events = {
      onClick: jest.fn(),
      onFocus: jest.fn(),
      onBlur: jest.fn(),
      onChange: jest.fn(),
    }
    const textToType = "typing into input"

    render(<TextInput name="test-input" label="Test Input" {...events} />)
    const received = await screen.findByLabelText("Test Input")

    await userEvent.type(received, textToType)
    userEvent.tab()

    expect(events.onClick).toHaveBeenCalledTimes(1)
    expect(events.onFocus).toHaveBeenCalledTimes(1)
    expect(events.onBlur).toHaveBeenCalledTimes(1)
    expect(events.onChange).toHaveBeenCalledTimes(textToType.length)
  })

  it("renders the helper text", async () => {
    const helperText = "this is helper text"

    render(
      <TextInput
        name="test-input"
        label="Test Input"
        helperText={helperText}
      />,
    )
    const receivedInput = await screen.findByLabelText("Test Input")
    const receivedHelperText = await screen.findByText(helperText)

    expect(receivedHelperText).toBeInTheDocument()
    expect(receivedInput).toHaveAttribute(
      "aria-describedby",
      receivedHelperText.id,
    )
  })

  it("renders correctly when hideLabel is true", async () => {
    render(<TextInput name="test-input" label="Test Input" hideLabel={true} />)
    const received = await screen.findByLabelText("Test Input")
    const receivedText = screen.queryByText("Test Input")

    expect(received).toBeInTheDocument()
    expect(received).toHaveAttribute("name", "test-input")
    expect(receivedText).not.toBeInTheDocument()
  })

  it("can not be typed into when disabled", async () => {
    const events = {
      onChange: jest.fn(),
      onBlur: jest.fn(),
      onFocus: jest.fn(),
      onClick: jest.fn(),
    }
    const textToType = "typing into input"

    render(
      <TextInput name="test-input" label="Test Input" disabled {...events} />,
    )
    const received = await screen.findByLabelText("Test Input")

    userEvent.type(received, textToType)
    userEvent.tab()

    expect(events.onClick).toHaveBeenCalledTimes(0)
    expect(events.onFocus).toHaveBeenCalledTimes(0)
    expect(events.onChange).toHaveBeenCalledTimes(0)
    expect(events.onBlur).toHaveBeenCalledTimes(0)
  })

  it("renders correctly when helperText is provided", async () => {
    const helperText = "this is helper text"

    render(
      <TextInput
        name="test-input"
        label="Test Input"
        helperText={helperText}
      />,
    )
    const receivedInput = await screen.findByLabelText("Test Input")
    const receivedHelperText = await screen.findByText(helperText)

    expect(receivedHelperText).toBeInTheDocument()
    expect(receivedInput).toHaveAttribute(
      "aria-describedby",
      receivedHelperText.id,
    )
  })
})
