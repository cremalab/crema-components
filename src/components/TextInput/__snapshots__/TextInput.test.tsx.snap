// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders the expected markup 1`] = `
<DocumentFragment>
  <div>
    <label
      for="test-input"
      id="test-input-label"
    >
      Test Input
    </label>
    <div>
      <input
        aria-describedby="test-input-helper-text"
        aria-label="Test Input"
        label="Test Input"
        name="test-input"
      />
    </div>
    <div
      id="test-input-helper-text"
    >
      Test Helper Text
    </div>
  </div>
</DocumentFragment>
`;
